File: Ring_Buffer/RingBuffer.cpp
#include "Precompile.h"
#include "RingBuffer.h"

#include <string.h>

RINGBUFFER::RINGBUFFER()
{
	buffer = new char[BUFSIZE];
	front = rear = 0;
}

RINGBUFFER::~RINGBUFFER()
{
	delete[] buffer;
}

int RINGBUFFER::GetUseSize()
{
	if (front > rear)
		return BUFSIZE - (front - rear);	// (BUFSIZE - 1) - (front - rear) + 1; 
	else
		return rear - front;
}

int RINGBUFFER::GetUnuseSize()
{
	if (front > rear)
		return front - rear - 1;
	else
		return (BUFSIZE - 1) - (rear - front);
}

bool RINGBUFFER::Enqueue(char *data, int size)
{
	int gap, unusing_data = GetUnuseSize();

	if ((rear + 1) % BUFSIZE == front || unusing_data < size)
		return false;

	rear++;
	if (rear == BUFSIZE)
		rear = 0;

	gap = BUFSIZE - rear;
	if (gap < size)
	{
		memcpy_s(buffer + rear, BUFSIZE, data, gap);
		memcpy_s(buffer, BUFSIZE, data + gap, size - gap);
		rear = size - gap - 1;
	}
	else
	{
		memcpy_s(buffer + rear, BUFSIZE, data, size);

		rear = (rear + size - 1);
		if (rear >= BUFSIZE)
			rear -= BUFSIZE;
	}

	return true;
}

bool RINGBUFFER::Dequeue(char *dest, int dest_size, int size)
{
	int gap, using_data = GetUseSize();

	if (rear == front || using_data < size)
		return false;

	front++;
	if (front == BUFSIZE)
		front = 0;

	gap = BUFSIZE - front;
	if (gap < size)
	{
		memcpy_s(dest, dest_size, buffer + front, gap);
		memcpy_s(dest + gap, dest_size, buffer, size - gap);
		front = size - gap - 1;
	}
	else
	{
		memcpy_s(dest, dest_size, buffer + front, size);

		front = (front + size - 1);
		if (front >= BUFSIZE)
			front -= BUFSIZE;
	}

	return true;
}

bool RINGBUFFER::Peek(char *dest, int dest_size, int size, int *return_size)
{
	int gap, temp, using_data = GetUseSize();

	if (rear == front || using_data < size)
		return false;

	temp = front + 1;
	if (temp == BUFSIZE)
		temp = 0;

	gap = BUFSIZE - temp;
	if (gap < size)
	{
		memcpy_s(dest, dest_size, buffer + temp, gap);
		size -= gap;	*return_size += gap;

		memcpy_s(dest + gap, dest_size, buffer, size);
		*return_size += size;
	}
	else
	{
		memcpy_s(dest, dest_size, buffer + temp, size);
		*return_size += size;
	}

	return true;
}

void RINGBUFFER::ClearBuffer()
{
	front = rear = 0;
}

char* RINGBUFFER::GetFrontPtr()
{
	int front_temp = front + 1;
	if (front_temp == BUFSIZE)
		front_temp = 0;

	return buffer + front_temp;
}

char* RINGBUFFER::GetRearPtr()
{
	int rear_temp = rear + 1;
	if (rear_temp == BUFSIZE)
		rear_temp = 0;

	return buffer + rear_temp;
}

void RINGBUFFER::MoveFront(int size)
{
	front += size;
	if (front >= BUFSIZE)
		front -= BUFSIZE;
}

void RINGBUFFER::MoveRear(int size)
{
	rear += size;
	if (rear >= BUFSIZE)
		rear -= BUFSIZE;
}

int RINGBUFFER::LinearRemainFrontSize()
{
	if (front == rear)
		return 0;
	else
	{
		int front_temp = front + 1;
		if (front_temp == BUFSIZE)	front_temp = 0;

		if (front_temp > rear)
			return BUFSIZE - front_temp;
		else
			return rear - front_temp + 1;
	}
}

int RINGBUFFER::LinearRemainRearSize()
{
	int rear_temp = rear + 1;
	if (rear_temp == BUFSIZE)	rear_temp = 0;

	if (rear_temp > front)
		return BUFSIZE - rear_temp;
	else
		return front - rear_temp;
}

File: C#_Library/RingBuffer.cs
using System;

namespace RINGBUFFER
{
    class RingBuffer
    {
        Byte[] m_buffer;
        Int32 m_front, m_rear;

        private enum Initial_Value : Int32
        {
            en_buffer_size = 100
        }

        public RingBuffer()
        {
            m_buffer = new byte[(Int32)Initial_Value.en_buffer_size];
            m_front = m_rear = 0;
        }


        /*
            버퍼에 데이터를 삽입 및 삭제하는 함수
        */
        unsafe public bool Enqueue(IntPtr p_input_data, Int32 p_size)
        {
            Int32 gap, t_front = m_front, t_rear = m_rear, unusing_data = GetUnuseSize();

            if ((t_rear + 1) % (Int32)Initial_Value.en_buffer_size == t_front || unusing_data < p_size)
                return false;

            t_rear++;
            if (t_rear == (Int32)Initial_Value.en_buffer_size)
                t_rear = 0;
            gap = (Int32)Initial_Value.en_buffer_size - t_rear;
            fixed (Byte* fix_buffer = m_buffer)
            {
                if (gap < p_size)
                {
                    switch (gap)
                    {
                        case 1: *(Byte*)(fix_buffer + t_rear)  = *(Byte*)p_input_data; break;
                        case 2: *(Int16*)(fix_buffer + t_rear) = *(Int16*)p_input_data; break;
                        case 4: *(Int32*)(fix_buffer + t_rear) = *(Int32*)p_input_data; break;
                        case 8: *(Int64*)(fix_buffer + t_rear) = *(Int64*)p_input_data; break;
                        default: System.Runtime.InteropServices.Marshal.Copy(p_input_data, m_buffer, t_rear, gap); break;
                    }

                    Byte* byte_store = (Byte*)(p_input_data.ToPointer());
                    byte_store += gap;

                    IntPtr intptr_store = (IntPtr)byte_store;

                    System.Runtime.InteropServices.Marshal.Copy(intptr_store, m_buffer, 0, p_size - gap);
                    m_rear = p_size - gap - 1;
                }
                else
                {
                    switch (p_size)
                    {
                        case 1: *(Byte*)(fix_buffer + t_rear)  = *(Byte*)p_input_data; break;
                        case 2: *(Int16*)(fix_buffer + t_rear) = *(Int16*)p_input_data; break;
                        case 4: *(Int32*)(fix_buffer + t_rear) = *(Int32*)p_input_data; break;
                        case 8: *(Int64*)(fix_buffer + t_rear) = *(Int64*)p_input_data; break;
                        default: System.Runtime.InteropServices.Marshal.Copy(p_input_data, m_buffer, t_rear, p_size); break;
                    }

                    if ((Int32)Initial_Value.en_buffer_size <= m_rear + p_size)
                        m_rear += p_size - (Int32)Initial_Value.en_buffer_size;
                    else
                        m_rear += p_size;
                }
            }

            return true;
        }

        unsafe public bool Dequeue(IntPtr p_output_data, Int32 p_size)
        {
            Byte* byte_store;
            Int32 gap, t_front = m_front, t_rear = m_rear, using_data = GetUseSize();

            if (t_rear == t_front || using_data < p_size)
                return false;

            t_front++;
            if (t_front == (Int32)Initial_Value.en_buffer_size)
                t_front = 0;

            gap = (Int32)Initial_Value.en_buffer_size - t_front;
            fixed (Byte* fix_buffer = m_buffer)
            {
                if (gap < p_size)
                {
                    switch (gap)
                    {
                        case 1: *(Byte*)p_output_data  = *(Byte*)(fix_buffer + t_front); break;
                        case 2: *(Int16*)p_output_data = *(Int16*)(fix_buffer + t_front); break;
                        case 4: *(Int32*)p_output_data = *(Int32*)(fix_buffer + t_front); break;
                        case 8: *(Int64*)p_output_data = *(Int64*)(fix_buffer + t_front); break;
                        default: System.Runtime.InteropServices.Marshal.Copy(m_buffer, t_front, p_output_data, gap); break;
                    }

                    byte_store = (Byte*)(p_output_data.ToPointer()); byte_store += gap;
                    IntPtr intptr_store = (IntPtr)byte_store;

                    System.Runtime.InteropServices.Marshal.Copy(m_buffer, 0, intptr_store, p_size - gap);
                    m_front = p_size - gap - 1;
                }
                else
                {
                    switch (p_size)
                    {
                        case 1: *(Byte*)p_output_data  = *(Byte*)(fix_buffer + t_front); break;
                        case 2: *(Int16*)p_output_data = *(Int16*)(fix_buffer + t_front); break;
                        case 4: *(Int32*)p_output_data = *(Int32*)(fix_buffer + t_front); break;
                        case 8: *(Int64*)p_output_data = *(Int64*)(fix_buffer + t_front); break;
                        default: System.Runtime.InteropServices.Marshal.Copy(m_buffer, t_front, p_output_data, p_size); break;
                    }

                    if ((Int32)Initial_Value.en_buffer_size <= m_front + p_size)
                        m_front += p_size - (Int32)Initial_Value.en_buffer_size;
                    else
                        m_front += p_size;
                }
            }

            return true;
        }

        unsafe public bool Peek(IntPtr p_output_data, Int32 p_size, ref Int32 p_return_size)
        {
            Byte* byte_store;
            Int32 gap, temp, t_front = m_front, t_rear = m_rear, using_data = GetUseSize();

            if (t_rear == t_front || using_data < p_size)
                return false;

            temp = t_front + 1;
            if (temp == (Int32)Initial_Value.en_buffer_size)
                temp = 0;

            gap = (Int32)Initial_Value.en_buffer_size - temp;
            fixed (Byte* fix_buffer = m_buffer)
            {
                if (gap < p_size)
                {
                    switch (gap)
                    {
                        case 1: *(Byte*)p_output_data  = *(Byte*)(fix_buffer + temp); break;
                        case 2: *(Int16*)p_output_data = *(Int16*)(fix_buffer + temp); break;
                        case 4: *(Int32*)p_output_data = *(Int32*)(fix_buffer + temp); break;
                        case 8: *(Int64*)p_output_data = *(Int64*)(fix_buffer + temp); break;
                        default: System.Runtime.InteropServices.Marshal.Copy(m_buffer, temp, p_output_data, gap); break;
                    }

                    byte_store = (Byte*)(p_output_data.ToPointer()); byte_store += gap;
                    IntPtr intptr_store = (IntPtr)byte_store;

                    System.Runtime.InteropServices.Marshal.Copy(m_buffer, 0, intptr_store, p_size - gap);
                }
                else
                {
                    switch (p_size)
                    {
                        case 1: *(Byte*)p_output_data  = *(Byte*)(fix_buffer + temp); break;
                        case 2: *(Int16*)p_output_data = *(Int16*)(fix_buffer + temp); break;
                        case 4: *(Int32*)p_output_data = *(Int32*)(fix_buffer + temp); break;
                        case 8: *(Int64*)p_output_data = *(Int64*)(fix_buffer + temp); break;
                        default: System.Runtime.InteropServices.Marshal.Copy(m_buffer, temp, p_output_data, p_size); break;
                    }
                }

                p_return_size += p_size;
            }

            return true;
        }


        /*
            Enqueue 및 Dequeue 함수 이외 외적으로 버퍼의 front 및 rear 정보가 변경되어야할 경우 호출되는 함수
        */
        public void MoveFront(Int32 p_size)
        {
            Int32 t_front = m_front;

            t_front += p_size;
            if (t_front >= (Int32)Initial_Value.en_buffer_size)
                t_front -= (Int32)Initial_Value.en_buffer_size;

            m_front = t_front;
        }

        public void MoveRear(Int32 p_size)
        {
            Int32 t_rear = m_rear;

            m_rear += p_size;
            if (m_rear >= (Int32)Initial_Value.en_buffer_size)
                m_rear -= (Int32)Initial_Value.en_buffer_size;

            m_rear = t_rear;
        }


        /*
            원형 큐의 전체 사용 및 미사용 크기 얻는 함수 
        */
        public Int32 GetUseSize()
        {
            Int32 t_front = m_front, t_rear = m_rear;

            if (t_front > t_rear)
                return (Int32)Initial_Value.en_buffer_size - (t_front - t_rear);    // (BUFSIZE - 1) - (front - rear) + 1; 
            else
                return t_rear - t_front;
        }

        public Int32 GetUnuseSize()
        {
            Int32 t_front = m_front, t_rear = m_rear;

            if (t_front > m_rear)
                return t_front - t_rear - 1;
            else
                return ((Int32)Initial_Value.en_buffer_size - 1) - (t_rear - t_front);
        }


        /*
            원형 큐의 각 위치에서 끊어지지 않은 영역 (선형 영역) 의 크기를 얻는 함수  
        */
        public Int32 LinearRemainFrontSize()
        {
            Int32 t_front = m_front, t_rear = m_rear;

            if (t_front == t_rear)
                return 0;
            else
            {
                Int32 front_temp = t_front + 1;
                if (front_temp == (Int32)Initial_Value.en_buffer_size) front_temp = 0;

                if (front_temp > t_rear)
                    return (Int32)Initial_Value.en_buffer_size - front_temp;
                else
                    return t_rear - front_temp + 1;
            }
        }

        public Int32 LinearRemainRearSize()
        {
            Int32 t_front = m_front, t_rear = m_rear;

            Int32 rear_temp = t_rear + 1;
            if (rear_temp == (Int32)Initial_Value.en_buffer_size) rear_temp = 0;

            if (rear_temp >= t_front)
                return (Int32)Initial_Value.en_buffer_size - rear_temp;
            else
                return t_front - rear_temp;
        }


        /*
            원형 큐의 각 위치에 대한 버퍼 주소를 얻는 함수       
        */
        unsafe public IntPtr GetBasicPtr()
        {
            fixed (Byte* fix_buffer = m_buffer)
                return (IntPtr)fix_buffer;
        }

        unsafe public IntPtr GetFrontPtr()
        {
            Int32 front_temp = m_front + 1;
            if (front_temp == (Int32)Initial_Value.en_buffer_size)
                front_temp = 0;

            fixed (Byte* fix_buffer = m_buffer)
            {
                return (IntPtr)(fix_buffer + front_temp);
            }
        }

        unsafe public IntPtr GetRearPtr()
        {
            Int32 rear_temp = m_rear + 1;
            if (rear_temp == (Int32)Initial_Value.en_buffer_size)
                rear_temp = 0;

            fixed (Byte* fix_buffer = m_buffer)
            {
                return (IntPtr)(fix_buffer + rear_temp);
            }
        }
    }
}